quilt login

sirspinach opened the pull request.

kklin commented on quiltctl/command/exec.go:
```diff
@@ -28,13 +28,31 @@ type Exec struct {
+// ExecHelper shares functionality between quilt commands similar to `quilt exec`, ie
+// quilt commands that execute a command on a remote docker container.
+type ExecHelper interface {
+	Init(ssh.Client, string, client.Getter, string, int, bool) int
```
I have a patch for refactoring the SSH package that might help, I'll think about it later.

sirspinach replied:
alright, i'll take out this interface

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,111 @@
+		"the private key to use to connect to the host")
+	flags.Usage = func() {
+		fmt.Println("usage: quilt login [-H=<daemon_host>] [-i=<private_key>] " +
+			"<stitch_id> <command>")
```
There's no `<command>` right?

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,111 @@
+			"The container is identified by the stitch ID produced by " +
+			"`quilt containers`.")
+		fmt.Println("For example, to get a shell in container 5 with a " +
+			"specific private key: quilt login -t -i ~/.ssh/quilt 5")
```
Don't need the `-t`

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	bestShell := findBestAvailableShell(shells, lCmd.helper)
+	cmd := fmt.Sprintf("%s", bestShell)
```
Why this `sprintf`?

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,111 @@
+			return shell
+		}
+	}
+	return "/bin/sh"
```
What if `/bin/sh` doesn't exist? I'd just add `/bin/sh` to the list of shells to try, and return an error if we reach this point.

ejj replied:
Wouldn't the most direct approach, btw, be to just try to run /bin/bash, if it fails try the next one, if that fails try the next one?

ejj replied:
Meh good point, not sure what's best then.  I do think its reasonable to assume that all containers that have /bin/bash have /bin/sh as well, maybe we stick with something like the current approach.

mmussomele replied:
why not do something like `echo $0` or `echo $SHELL`?

sirspinach replied:
Not sure entirely what you mean, @mmussomele. If you're suggesting that we read `$SHELL` on the quilt client to use as a candidate shell, that sounds good to me.

mmussomele replied:
Yea pretty much

kklin replied:
Sounds complicated to me. At the very least, I think that'd be a feature that gets introduced in a later commit. The initial merge should just be trying from a list of shells.

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,111 @@
+// installed, findBestAvailableShell returns "/bin/sh".
+func findBestAvailableShell(shells []string, helper ExecHelper) string {
+	for _, shell := range shells {
+		log.Debug("try shell ", shell)
```
"trying"

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,111 @@
+func findBestAvailableShell(shells []string, helper ExecHelper) string {
+	for _, shell := range shells {
+		log.Debug("try shell ", shell)
+		checkCmd := fmt.Sprintf("/bin/sh -c 'hash %s 2>/dev/null; echo $?'",
```
What if `sh` doesn't exist? Can't we just invoke `hash` directly? And why can't we just check the exit code directly?

sirspinach replied:
`hash` is a shell built-in, so it needs to run within a shell.
```bash
shwangster@aetherboard ~/go/src/github.com/NetSys/quilt
 % docker run --rm ubuntu hash
docker: Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"exec: \\\"hash\\\": executable file not found in $PATH\"\n".

shwangster@aetherboard ~/go/src/github.com/NetSys/quilt
 % which hash
hash: shell built-in command
```

kklin replied:
Meh, then I don't think `hash` is the right approach.

sirspinach replied:
Don't you think it's a safe assumption to make that /bin/sh exists on a unix system?

kklin replied:
I'm not sure. If it really is true that all containers have `sh` then a comment at least would be good.

sirspinach commented:
Removed ExecHelper interface and replaced it with a execHelper function. 
Check shells existence using exit codes rather than stdout.
Check existence of $SHELL, as suggested by @mmussomele.

sirspinach commented:
I'm waiting until @kklin's SSH refactor before I write tests for login.go, because login.go will probably become a lot shorter then.

kklin commented:
Ah, I didn't realize you were blocking on that @sirspinach. I'll prioritize it.

sirspinach commented:
Ready for further review.

kklin commented on quiltctl/ssh/ssh.go:
```diff
@@ -2,12 +2,11 @@ package ssh
+	Close() error
 }
+
+// Getter is used to retrieve an Client
```
"a Client"

kklin commented on quiltctl/command/exec.go:
```diff
@@ -118,6 +119,12 @@ func (eCmd *Exec) Run() int {
+		if exitErr, ok := err.(*golangSSH.ExitError); ok {
+			log.WithError(err).Debug(
+				"SSH command returned a nonzero exit code")
+			return exitErr.ExitStatus()
```
Ah, this is much better.

kklin commented on quiltctl/command/exec.go:
```diff
@@ -97,7 +97,6 @@ func (eCmd *Exec) Run() int {
 		log.WithError(err).Error("Error getting container client")
 		return 1
 	}
-	defer containerClient.Close()
```
Did you mean to delete this line?

kklin commented on quiltctl/command/exec.go:
```diff
@@ -112,19 +111,28 @@ func (eCmd *Exec) Run() int {
 	}
 	defer sshClient.Close()
 
+	if exitCode := execHelper(eCmd.command, sshClient, container.DockerID,
```
Just return the exit code

kklin commented on quiltctl/command/exec.go:
```diff
@@ -112,19 +111,28 @@ func (eCmd *Exec) Run() int {
+	return 0
+}
+
+func execHelper(cmd string, sshClient ssh.Client, dockerID string, requestPTY bool) int {
```
I'd prefer the order of arguments to be: `(sshClient, dockerID, cmd, requestPTY)`

kklin commented on quiltctl/command/exec.go:
```diff
@@ -112,19 +111,28 @@ func (eCmd *Exec) Run() int {
 				"SSH command returned a nonzero exit code")
 			return exitErr.ExitStatus()
 		}
-
```
You add this new line in a previous commit, then remove it here I think. To make the diff cleaner I think you should pick one of the two and just stick with it.

kklin commented on quiltctl/command/exec_test.go:
```diff
@@ -155,3 +156,32 @@ func checkExecParsing(t *testing.T, args []string, expArgs Exec, expErr error) {
+	//SSH ExitError
+	mockSSHClient := new(testutils.MockSSHClient)
+	mockSSHClient.On("Run", true, "docker exec -it id123 echo hello").
+		Return(&golangSSH.ExitError{}).Once()
```
Could you make the exit code non-zero? Makes it more obvious what it's testing.

sirspinach replied:
Okay, but this will require editing an unexported variable in the `ssh.Waitmsg` struct.

sirspinach replied:
nvm, `reflect` doesn't allow me to change an unexported struct field.

sirspinach replied:
Gonna try making an interface.

kklin replied:
Meh... seems like a huge amount of effort.

kklin replied:
I'm fine leaving it as is

kklin commented on quiltctl/command/exec_test.go:
```diff
@@ -155,3 +156,32 @@ func checkExecParsing(t *testing.T, args []string, expArgs Exec, expErr error) {
 }
+
+func TestExecHelperErrors(t *testing.T) {
+	//SSH ExitError
```
Space after `//`.

kklin commented on quiltctl/command/exec_test.go:
```diff
@@ -155,3 +156,32 @@ func checkExecParsing(t *testing.T, args []string, expArgs Exec, expErr error) {
+	assert.Equal(t, 0, exitCode)
+	mockSSHClient.AssertExpectations(t)
+
+	//Other SSH error
```
Same here re space

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,157 @@
+	defer sshClient.Close()
+
+	bestShell := findBestAvailableShell(shells, sshClient, container.DockerID)
+	if bestShell == "" {
```
Eh, seems hacky. I'd prefer having `findBestAvailableShell` return an error.

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,157 @@
+// docker container, first trying shells listed in `shells`, then /bin/sh.
+//
+// If findBestAvailableShell fails to find a shell, then it returns the empty string.
+// The current implementation uses /bin/sh to test whether other shells exist, and will
```
I'm still iffy about depending on `/bin/sh`.. But this is definitely better than having nothing, and if it ends up being a bad assumption we can revisit this implementation.

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+var tryShell = func(shell string, sshClient ssh.Client, dockerID string) bool {
+	log.Debug("trying shell ", shell)
```
Nit - I prefer making the caller log as much as possible. So I'd move the `log`s in this func to inside the loop.

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,157 @@
+var tryShell = func(shell string, sshClient ssh.Client, dockerID string) bool {
+	log.Debug("trying shell ", shell)
+
+	checkCmd := fmt.Sprintf("/bin/sh -c 'hash %s 2>/dev/null'", shell)
```
Does this redirect `stdout`?

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,157 @@
+	log.Debug("trying shell ", shell)
+
+	checkCmd := fmt.Sprintf("/bin/sh -c 'hash %s 2>/dev/null'", shell)
+	errCode := execHelper(checkCmd, sshClient, dockerID, false)
```
Once you make the `log` changes, simply `return errCode == 0`.

kklin commented on quiltctl/command/login_test.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+func TestLoginDefaultToSh(t *testing.T) {
+	origTryShell := tryShell
```
Eh, I don't think we need to set that func back

sirspinach replied:
`TestLoginTryShellSuccess` also uses the original `tryShell()`, and failed after I removed this `defer func() { tryShell = origTryShell }()`.

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,157 @@
+	return "/bin/sh"
+}
+
+var tryShell = func(shell string, sshClient ssh.Client, dockerID string) bool {
```
I'd prefer the order to be `sshClient, dockerID, shell`

sirspinach replied:
For consistency, I'm also changing the order of findBestAvailableShell to `sshClient, dockerID, shells`

kklin commented on quiltctl/command/login.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"errors"
```
I'm not a big fan of this feature -- it seems to just complicate things without much benefit. I don't see any reason why you'd expect your local machine's shell to have any relation to the container's.

(cc @mmussomele @ejj @LuiseV)

sirspinach replied:
In case this is relevant, docker run and exec both return the exit code of the commands they run.
 ```
% docker run --rm ubuntu bash -c false; echo $?
1
```

sirspinach replied:
Don't feel strongly about whether or not to include this.

An argument for this patch: a user who is using `zsh` would most likely prefer `zsh` over others shells when he or she logs into a container. Though it's unlikely that anything other than bash, ash, and sh would be automatically installed in a container.

kklin replied:
Yeah, and that user won't have their dotfiles in place.

sirspinach commented:
ready for further review.

kklin commented on quiltctl/command/exec_test.go:
```diff
@@ -14,6 +15,18 @@ import (
+}
+
+func (err mockExitError) Error() string {
+	return fmt.Sprint("Exitted with error code: ", err.exitCode)
```
"Exited"

kklin commented on quiltctl/command/exec_test.go:
```diff
@@ -20,7 +20,7 @@ type mockExitError struct {
 
 func (err mockExitError) Error() string {
-	return fmt.Sprint("Exitted with error code: ", err.exitCode)
+	return fmt.Sprint("exitted with error code: ", err.exitCode)
```
Should be changed in the previous commit

kklin commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+// Run logs into a shell on the remote container.
+func (lCmd *Login) Run() int {
```
Could you add some coverage for this method?

sirspinach commented on quiltctl/command/exec.go:
```diff
@@ -31,6 +30,13 @@ type Exec struct {
 
+// exitError is an interface to "golang.org/x/crypto/ssh".ExitError that allows for
+// mocking in unit tests.
+type exitError interface {
```
exitError should have been added in the previous commit.

sirspinach commented:
ready for more review.

ejj commented on quiltctl/ssh/native.go:
```diff
@@ -86,10 +86,11 @@ func (c NativeClient) Run(requestPTY bool, command string) error {
 	session.Stderr = os.Stderr
 
 	if requestPTY {
+		session.Stdin = os.Stdin
```
I don't see the point of this patch, why not just do it in the previous patch when you introduce the bug?

sirspinach replied:
This bug was around even before the SSH refactor: https://github.com/NetSys/quilt/pull/639/commits/bc9084794c5597be3e89fa6a11796391a572d612#diff-0a77957168c83c3e168d5673d6900f35L111

ejj commented:
@kklin, @sirspinach, what's the status on this one?  Did we come to a conclusion on the last commit?  I don't have enough bash knowledge to have an intelligent opinion on it.

kklin commented:
Let's ditch it. I think there was miscommunication on what feature @mmussomele wanted (it was an implementation suggestion, not a feature request), I don't like it, and @sirspinach isn't attached to it either.

Certainly shouldn't block the other stuff getting merged.

ejj commented:
kk @sirspinach, could you rebase and remove the last commit?  I'll review it then.

sirspinach commented:
rebased

On Jan 9, 2017 3:10 PM, "Ethan J. Jackson" <notifications@github.com> wrote:

> kk @sirspinach <https://github.com/sirspinach>, could you rebase and
> remove the last commit? I'll review it then.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NetSys/quilt/pull/639#issuecomment-271437447>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABq3MyYvPyfgzWDbc_uEnlL-Pq3dbFxyks5rQr5ZgaJpZM4LJ0ek>
> .
>

ejj commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,153 @@
+	flags.StringVar(&lCmd.privateKey, "i", "",
+		"the private key to use to connect to the host")
+	flags.Usage = func() {
+		fmt.Println("usage: quilt login [-H=<daemon_host>] [-i=<private_key>] " +
```
I think you should pull this stuff into a global variable and use the triple backticks.  It would make the code a bit easier to read, and it wouldn't count against our code coverage.

sirspinach replied:
fixed

ejj commented on quiltctl/ssh/ssh.go:
```diff
@@ -2,12 +2,11 @@ package ssh
-	Run(string) error
-	// Disconnect closes the SSH connection.
-	Disconnect() error
+	Run(bool, string) error
```
I think I'd prefer a newline under Run

sirspinach replied:
done

ejj commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,153 @@
+// If findBestAvailableShell fails to find a shell, then it returns the empty string.
+// The current implementation uses /bin/sh to test whether other shells exist, and will
+// fail if /bin/sh is not found on the remote container.
+func findBestAvailableShell(sshClient ssh.Client, dockerID string, shells []string) (
```
This function seems a bit odd to me.  Why make the special case for `/bin/sh`?  Seems to me you could just add it to the end of the global `shells` variable and detect it in the tryShell loop.  If none of them works, it could just `log.Fatal`

ejj replied:
Actually here's my biggest question.  Why can't we just loop through the `shells` variable calling `execHelper` on each one.  If it works, great, we've logged in.  If it doesn't work then we move on to the next one.

sirspinach replied:
Goes back to this discussion: https://github.com/NetSys/quilt/pull/639#discussion_r91862819
"docker logs scary messages in journalctl -u docker when you try to exec something that doesn't exit"

ejj replied:
Ah I see.

Still, could we make this a bit simpler?  So for example you could imagine we could try `bash -c pwd`, `ash -c pwd`, `sh -c pwd` for each shell to check if it exists?  That way you could get rid of the special handling for sh?

sirspinach replied:
All of those commands result in journalctl errors when the shell doesn't exist. :(

sirspinach replied:
If we want to avoid journalctl logs, then we cannot `docker exec` any command that does not exist. So in this approach I assumed `sh`'s existence and then used that command to test the other shells.

kklin replied:
FWIW docker assumes `sh` exists when using the "_shell_ form" of CMD https://docs.docker.com/engine/reference/builder/#/cmd.

kklin replied:
Another option is to check the filesystem directly..

ejj commented:
K done with my review, couple of nits, but my biggest question is the very last comment.

sirspinach commented on quiltctl/command/login.go:
```diff
@@ -0,0 +1,109 @@
+		"the private key to use to connect to the host")
+
+	flags.Usage = func() {
+		fmt.Println(loginUsage)
```
I think that in a later patch, we could do away with this `flags.Usage = func() {` part (and therefore the lack of coverage in usage messages) by adding a `usageMsg` argument to `lCmd.common.InstallFlags(flags)`.

sirspinach commented:
Changed login to use `sh` only.
